discoverd/client/deployment.go

```diff
@@ -0,0 +1,195 @@
+package discoverd
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type DeploymentState string
+
+const (
+	DeploymentStatePerforming DeploymentState = "performing"
+	DeploymentStateDone       DeploymentState = "done"
+)
+
+func NewDeployment(service string) (*Deployment, error) {
+	s := NewService(service)
+	events := make(chan *Event)
+	stream, err := s.Watch(events)
+	if err != nil {
+		return nil, err
+	}
+	return &Deployment{s, events, stream}, nil
+}
+
+// Deployment is a wrapper around service metadata for marking jobs as either
+// performing a deployment or done deploying
+type Deployment struct {
+	service Service
+	events  chan *Event
+	stream  stream.Stream
+}
+
+const deploymentTimeout = time.Minute
+
+// MarkPerforming marks the given address as performing in the service metadata,
+// ensuring there is only one address marked as performing at any given time
+// (waiting for a ServiceMeta event and retrying if there is already an
+// address performing).
+func (d *Deployment) MarkPerforming(addr string) error {
+outer:
+	for {
+		meta, err := d.service.GetMeta()
+		if err != nil {
+			return err
+		}
+
+		states, err := d.decode(meta)
+		if err != nil {
+			return err
+		}
+
+		performing := ""
+		for a, state := range states {
+			if a == addr {
+				// already marked as performing, nothing to do
+				return nil
+			}
+
+			if state == DeploymentStatePerforming {
+				performing = a
+				break
+			}
+		}
+
+		// if another address is performing, wait for a ServiceMeta
+		// event then try again
+		if performing != "" {
+			for {
+				select {
+				case event, ok := <-d.events:
+					if !ok {
+						return fmt.Errorf("service stream closed unexpectedly: %s", d.stream.Err())
+					}
+					if event.Kind == EventKindServiceMeta {
+						continue outer
+					}
+				case <-time.After(deploymentTimeout):
+					return fmt.Errorf("timed out waiting for %s to finish performing", performing)
+				}
+			}
+		}
+
+		// no performing addresses, attempt to mark addr
+		states[addr] = DeploymentStatePerforming
+
+		data, err := json.Marshal(states)
+		if err != nil {
+			return err
+		}
+		meta.Data = data
+
+		if err := d.service.SetMeta(meta); err == nil {
+			return nil
+		}
```

Comment:
This loop should probably have logging (especially the error here, but probably most of the code paths)